File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/UpperEvaluator.java
package org.opencds.cqf.cql.engine.elm.visiting;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;

/*
Upper(argument String) String

The Upper operator returns the upper case of its argument.
If the argument is null, the result is null.
*/

public class UpperEvaluator {

    public static Object upper(Object operand) {
        if (operand == null) {
            return null;
        }

        if (operand instanceof String) {
            return ((String) operand).toUpperCase();
        }

        throw new InvalidOperatorArgument(
                "Upper(String)",
                String.format("Upper(%s)", operand.getClass().getName())
        );
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/UpperEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.CqlType
import org.opencds.cqf.cql.engine.runtime.String
import org.opencds.cqf.cql.engine.runtime.toCqlString

/*
Upper(argument String) String

The Upper operator returns the upper case of its argument.
If the argument is null, the result is null.
*/
object UpperEvaluator {
    @JvmStatic
    fun upper(operand: CqlType?): String? {
        if (operand == null) {
            return null
        }

        if (operand is String) {
            return operand.value.uppercase().toCqlString()
        }

        throw InvalidOperatorArgument("Upper(String)", "Upper(${operand.typeAsString})")
    }
}
